File: 2015/hs/src/day02.hs
Type Box = [Char]

split :: Box -> [Int]
split "" = []
split s = fp : rest
  where parts = break (=='x') s
        fp = read (fst parts)::Int
        rest = split $ drop 1 $ snd parts

wrapBox :: [Int] -> Int
wrapBox b@(l:w:h:[]) = area + minimum b
  where area = sum d * 2
        d = zipWith (*) [l, w, h] [w, h, l]

part1 :: (Foldable t, Functor t) => t Box -> Int
part1 ls = sum $ fmap (wrapBox . split) ls

-- part1 <$> lines <$> getLine
-- part1 <$> lines <$> readFile "02.input"

File: 2020/js/day02.js
const read = require('./read.js')
const fn = read('02')

const line = s => {
  const parts = s.split(' ')
  return {
    r: parts[0].split('-').map(Number),
    ch: parts[1][0],
    pw: parts[2]
  }
}

const count = (str, c) => str
  .split('')
  .filter(d => d === c)
  .length

const part1 = fn.map(line).map(l => {
  const { r, ch, pw } = l
  const c = count(pw, ch)
  return c >= r[0] && c <= r[1]
}).filter(d => !!d).length

const part2 = fn.map(line).map(l => {
  const { r, ch, pw } = l
  const p1 = pw[r[0]-1] === ch
  const p2 = pw[r[1]-1] === ch
  return (p1 && !p2) || (!p1 && p2)
}).filter(d => !!d).length

console.log(part1, part2)